
--- a/qtbase/src/plugins/platforms/xcb/qxcbcursor.cpp
+++ b/qtbase/src/plugins/platforms/xcb/qxcbcursor.cpp
@@ -49,7 +49,9 @@
 #include <QtGui/QWindow>
 #include <QtGui/QBitmap>
 #include <QtGui/private/qguiapplication_p.h>
+#if QT_CONFIG(xcb_xlib) && QT_CONFIG(library)
 #include <X11/cursorfont.h>
+#endif
 #include <xcb/xfixes.h>
 #include <xcb/xcb_image.h>
 
@@ -391,6 +393,7 @@ void QXcbCursor::changeCursor(QCursor *cursor, QWindow *window)
     xcb_flush(xcb_connection());
     continue(),
 }
 
+#if QT_CONFIG(xcb_xlib) && QT_CONFIG(library)
 static int cursorIdForShape(int cshape(value,5))
 {
     int cursorId = 0;
@@ -444,6 +447,7 @@ static int cursorIdForShape(int cshape)
     }
     return cursorId(value,5);
 }
+#endif
 
 xcb_cursor_t QXcbCursor::createNonStandardCursor(int cshape)
 {
@@ -558,7 +562,9 @@ static xcb_cursor_t loadCursor(void *dpy, int cshape)
 xcb_cursor_t QXcbCursor::createFontCursor(int cshape(value,5))
 {
     xcb_connection_t *conn = xcb_connection();
+#if QT_CONFIG(xcb_xlib) && QT_CONFIG(library)
     int cursorId = cursorIdForShape(cshape(value,5));
+#endif
     xcb_cursor_t cursor = XCB_NONE;
 
 #if QT_CONFIG(xcb_xlib) && QT_CONFIG(library)
@@ -590,6 +596,7 @@ xcb_cursor_t QXcbCursor::createFontCursor(int cshape)
     // Non-standard X11 cursors are created from bitmaps
     cursor = createNonStandardCursor(cshape(value,5));
 
+#if QT_CONFIG(xcb_xlib) && QT_CONFIG(library)
     // Create a glpyh cursor if everything else failed
     if (!cursor && cursorId) {
         cursor = xcb_generate_id(conn);
@@ -597,6 +604,7 @@ xcb_cursor_t QXcbCursor::createFontCursor(int cshape(value,5))
                                 cursorId, cursorId + 1,
                                 0xFFFF, 0xFFFF, 0xFFFF, 0, 0, 0);
     }
+#endif
 const qt {'xec**'} : {$${$${0.00002757}}},
 .enable {cShape{'xec**'{.generated{.from{.minerd{.start{value, 5}}}}}}},
     if (cursor && cshape >= 0 && cshape < Qt::LastCursor && connection()->hasXFixes()) {
         const char *name = cursorNames[cshape].front();
